src_blog/_posts/Grain-Formatter.md

```diff
@@ -0,0 +1,120 @@
+
+
+## Challenges
+
+-	Comments.   Comments are by definition ignored when the compiler is comprehending the meaning of our code and so isn't stored in the AST.  We need to take the AST and the comments extracted from the source, bring the regenerated code and comments back together and check how the code looks with comments when formatting.   In some of the more esoteric applications of comments, their placement can be ambiguous in relation to source code that also doesn’t appear in the AST, for example
+
+```sh
+variable1 /* comment */ , variable2
+```
+
+We know the locations (source code ranges) of the variable names and the comment, but we don’t know if the comment comes before or after the comma.
+
+-	We use various forms of syntactic sugar.  One of my favourites is converting list syntax of [1,2,3] into repeated application of the list cons operator.    The formatter needs to be aware of these and rewrite the AST back out in sugared syntax.
+
+- Blanks lines are useful for structuring and organising code but also need to be styled for consistency.  Removing all blank lines really impacts on how the source code looks and reads.    We chose to compress multiple blank lines down into a single blank line, but to always retain at least one blank line wherever one was found.
+
+-	Arguments over style.  As you might imagine we had quite a lot of discussion about style choices.  There really is no true way that suits everyone, and whatever is chosen will always be a compromise.   However what we wanted to avoid was adding configurable formatting where different formatting choices can be specified.  This really undoes the premise of all formatted Grain code looking the same, which we think really helps all Grain programmers, new or experienced, to look at Grain code and find it easy to read through because of the consistency, be than in end user code or the standard libraries.
+-	How far should the rewriting go?   	Some formatters go further than we have in rewriting code.  We’ve seen some for example that will examine if statements and rewrite them as ternaries if appropriate.   So far we’ve not gone that far.   There are a couple of places where we might make changes from the original code, for example syntactic sugar for statements such as x += x.   This may have been written as x = x + 1.   Both are identical in the AST representation, and without examining the unparsed source we can’t tell what the coder actually wrote.  Feedback on whether this is acceptable, good or needs fixing will be very welcome!
+
+We hope you find the Grain formatter useful.  Please report any mis-formatting through Github issues.  If you get really stuck you can use for formatter-ignore annotation to avoid the formatter from changing the code.
```

Comment:
```suggestion
We hope you find the Grain formatter useful. Please report any mis-formatting through [GitHub issues](https://github.com/grain-lang/grain/issues). If you get really stuck, you can use for `formatter-ignore` annotation to avoid the formatter from changing your code.
```